File: 백준/Gold/15686. 치킨 배달/치킨 배달.py
from collections import deque
import sys

N, M = map(int, input().split())
house = []
chicken = []
open_chicken = [[False] * N for _ in range(N)]
moves = [(-1, 0), (1, 0), (0, -1), (0, 1)]
min_chicken_distance = sys.maxsize

# 입력
for r in range(N):
    row = list(map(int, input().split()))
    for c in range(N):
        if row[c] == 1:
            house.append((r, c))
        elif row[c] == 2:
            chicken.append((r, c))

def make_combination(depth, start):
    global min_chicken_distance
    
    if depth == M:
        # 치킨 거리 계산
        distance = 0
        for h in house:
            distance += calc_chicken_distance(h)
        if min_chicken_distance > distance:
            min_chicken_distance = distance
        return
    
    for i in range(start, len(chicken)):
        r = chicken[i][0]
        c = chicken[i][1]
        open_chicken[r][c] = True
        make_combination(depth + 1, i + 1)
        open_chicken[r][c] = False

def calc_chicken_distance(start):
    distance = sys.maxsize
    for r, c in chicken:
        if open_chicken[r][c]:
            temp = abs(start[0]-r) + abs(start[1]-c)
            if distance > temp:
                distance = temp
    return distance

make_combination(0, 0)
print(min_chicken_distance)

File: 백준/Gold/15686. 치킨 배달/치킨 배달.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static StringTokenizer tokens;
    private static StringBuilder sb = new StringBuilder();

    private static int N;
    private static int M;

    private static ArrayList<int[]> house = new ArrayList<>();
    private static ArrayList<int[]> chicken = new ArrayList<>();
    private static boolean[][] openChicken;
    private static int minChickenDistance = Integer.MAX_VALUE;

    private static int calcChickenDistance(int[] start) {
        int distance = Integer.MAX_VALUE;

        for (int[] coord : chicken) {
            int r = coord[0];
            int c = coord[1];
            if (openChicken[r][c]) {
                int temp = Math.abs(start[0] - r) + Math.abs(start[1] - c);
                distance = Math.min(distance, temp);
            }
        }

        return distance;
    }

    private static void makeCombination(int depth, int start) {
        if (depth == M) {
            int distance = 0;
            for (int[] h : house) {
                distance += calcChickenDistance(h);
            }
            minChickenDistance = Math.min(minChickenDistance, distance);
            return;
        }

        for (int i = start; i < chicken.size(); i++) {
            int r = chicken.get(i)[0];
            int c = chicken.get(i)[1];
            openChicken[r][c] = true;
            makeCombination(depth + 1, i + 1);
            openChicken[r][c] = false;
        }
    }

    private static void init() throws IOException {
        tokens = new StringTokenizer(br.readLine());
        N = Integer.parseInt(tokens.nextToken());
        M = Integer.parseInt(tokens.nextToken());
        openChicken = new boolean[N][N];

        for (int i = 0; i < N; i++) {
            tokens = new StringTokenizer(br.readLine());
            for (int j = 0; j < N; j++) {
                int temp = Integer.parseInt(tokens.nextToken());
                if (temp == 1) {
                    house.add(new int[]{i, j});
                }
                if (temp == 2) {
                    chicken.add(new int[]{i, j});
                }
            }
        }
    }

    public static void main(String[] args) throws IOException {
        init();
        makeCombination(0, 0);
        System.out.println(minChickenDistance);
    }
}